### ISSUE TITLE
Renaming Folders over more than two nested levels fails

### ISSUE THREAD

**Issue by sb-relaxt-at** [silverstripe/silverstripe-framework#4993]
Best shown by a little example to reproduce the issue: 

- Create a new folder named "1" on the root level (via the FileAdmin)
- Create a new folder named "2" inside of "1"
- Create a new folder named "3" inside of "2"

Finally we got 1 > 2 > 3. Renaming folder "1" to "1b" causes an error and an inconsistent db state. This happens also if "3" is a file and not a folder (although with a different error message). The error occurs in File#updateFilesystem and/or Folder#updateFilesystem. I was able to reproduce it in 3.2.1 as well as in 3.3.0-rc1.

**Comment by tractorcow** [comment]
In 4.0, folders won't be backed (necessarily) by physical local directories anymore.

Still, it needs to be solved in 3.x. :)

What is the error message? that would be helpful to include in your report. ;)

**Comment by sb-relaxt-at** [comment]
Oh yeah, we are looking forward to that :-).

Here we go (3.2.1, the described procedure with nested Folders):

```
ERROR [Warning]: rename(/.../silverstripe/3.2/assets/1/2/3/,/.../silverstripe/3.2/assets/3/) [<a href='function.rename'>function.rename</a>]: No such file or directory
IN POST /.../silverstripe/3.2/admin/assets/EditForm/field/File/item/24/ItemEditForm
Line 593 in /.../silverstripe/3.2/framework/filesystem/File.php

Source
======
  584: 				// Check that target directory (not the file itself) exists.
  585: 				// Only check if we're dealing with a file, otherwise the folder will need to be created
  586: 				if(!file_exists(dirname($pathAfterAbs))) {
  587: 					throw new Exception("Cannot move $pathBeforeAbs to $pathAfterAbs - Directory " .
       dirname($pathAfter)
  588: 						. " doesn't exist");
  589: 				}
  590: 			}
  591: 
  592: 			// Rename file or folder
* 593: 			$success = rename($pathBeforeAbs, $pathAfterAbs);
  594: 			if(!$success) throw new Exception("Cannot move $pathBeforeAbs to $pathAfterAbs");
  595: 		}
  596: 
  597: 
  598: 		// Update any database references
  599: 		$this->updateLinks($pathBefore, $pathAfter);

Trace
=====
rename(/.../silverstripe/3.2/assets/1/2/3/,/.../silverstripe/3.2/assets/3/)
File.php:593

File->updateFilesystem()
Folder.php:434

Folder->updateFilesystem()
Folder.php:440

Folder->updateFilesystem()
Folder.php:440

Folder->updateFilesystem()
File.php:533

File->onAfterWrite()
DataObject.php:1393

DataObject->write()
GridFieldDetailForm.php:547

GridFieldDetailForm_ItemRequest->doSave(Array,Form,SS_HTTPRequest)
Form.php:462

Form->httpSubmission(SS_HTTPRequest)
RequestHandler.php:288

RequestHandler->handleAction(SS_HTTPRequest,httpSubmission)
RequestHandler.php:200

RequestHandler->handleRequest(SS_HTTPRequest,DataModel)
RequestHandler.php:222

RequestHandler->handleRequest(SS_HTTPRequest,DataModel)
GridFieldDetailForm.php:102

GridFieldDetailForm->handleItem(GridField,SS_HTTPRequest)
GridField.php:963

GridField->handleRequest(SS_HTTPRequest,DataModel)
RequestHandler.php:222

RequestHandler->handleRequest(SS_HTTPRequest,DataModel)
RequestHandler.php:222

RequestHandler->handleRequest(SS_HTTPRequest,DataModel)
Controller.php:157

Controller->handleRequest(SS_HTTPRequest,DataModel)
LeftAndMain.php:451

LeftAndMain->handleRequest(SS_HTTPRequest,DataModel)
AdminRootController.php:93

AdminRootController->handleRequest(SS_HTTPRequest,DataModel)
Director.php:385

Director::handleRequest(SS_HTTPRequest,Session,DataModel)
Director.php:149

Director::direct(/admin/assets/EditForm/field/File/item/24/ItemEditForm,DataModel)
main.php:184

ERROR [User Error]: Uncaught Exception: Cannot move /.../silverstripe/3.2/assets/1/2/3/ to /.../silverstripe/3.2/assets/3/
IN POST /.../silverstripe/3.2/admin/assets/EditForm/field/File/item/24/ItemEditForm
Line 594 in /.../silverstripe/3.2/framework/filesystem/File.php

Source
======
  585: 				// Only check if we're dealing with a file, otherwise the folder will need to be created
  586: 				if(!file_exists(dirname($pathAfterAbs))) {
  587: 					throw new Exception("Cannot move $pathBeforeAbs to $pathAfterAbs - Directory " .
       dirname($pathAfter)
  588: 						. " doesn't exist");
  589: 				}
  590: 			}
  591: 
  592: 			// Rename file or folder
  593: 			$success = rename($pathBeforeAbs, $pathAfterAbs);
* 594: 			if(!$success) throw new Exception("Cannot move $pathBeforeAbs to $pathAfterAbs");
  595: 		}
  596: 
  597: 
  598: 		// Update any database references
  599: 		$this->updateLinks($pathBefore, $pathAfter);
  600: 	}

Trace
=====
File->updateFilesystem()
Folder.php:434

Folder->updateFilesystem()
Folder.php:440

Folder->updateFilesystem()
Folder.php:440

Folder->updateFilesystem()
File.php:533

File->onAfterWrite()
DataObject.php:1393

DataObject->write()
GridFieldDetailForm.php:547

GridFieldDetailForm_ItemRequest->doSave(Array,Form,SS_HTTPRequest)
Form.php:462

Form->httpSubmission(SS_HTTPRequest)
RequestHandler.php:288

RequestHandler->handleAction(SS_HTTPRequest,httpSubmission)
RequestHandler.php:200

RequestHandler->handleRequest(SS_HTTPRequest,DataModel)
RequestHandler.php:222

RequestHandler->handleRequest(SS_HTTPRequest,DataModel)
GridFieldDetailForm.php:102

GridFieldDetailForm->handleItem(GridField,SS_HTTPRequest)
GridField.php:963

GridField->handleRequest(SS_HTTPRequest,DataModel)
RequestHandler.php:222

RequestHandler->handleRequest(SS_HTTPRequest,DataModel)
RequestHandler.php:222

RequestHandler->handleRequest(SS_HTTPRequest,DataModel)
Controller.php:157

Controller->handleRequest(SS_HTTPRequest,DataModel)
LeftAndMain.php:451

LeftAndMain->handleRequest(SS_HTTPRequest,DataModel)
AdminRootController.php:93

AdminRootController->handleRequest(SS_HTTPRequest,DataModel)
Director.php:385

Director::handleRequest(SS_HTTPRequest,Session,DataModel)
Director.php:149

Director::direct(/admin/assets/EditForm/field/File/item/24/ItemEditForm,DataModel)
main.php:184


```

**Comment by tractorcow** [comment]
Oh, that is useful! Thanks for getting back with that.

It's just dumb logic in Folder::updateFilesystem. It's invoked in File::onAfterWrite() , but it CHANGES it's local fields (e.g. Filename) without writing them to the database.

Then when it iterates over children, all those child records are looking at the DB written version of the parent folder, which of course has the wrong Filename value (since the correct isn't saved to the DB yet).

What a mess. =O

**Comment by tractorcow** [comment]
I toyed with the idea of making folders NOT subclasses of File, but that would require a bit of refactoring to make AssetAdmin respect it properly.

**Comment by sb-relaxt-at** [comment]
You are welcome :-). I see, I assumed something like that. I am not 100% certain, but I believe this bug has been introduced somewhere after 3.2.0-beta1.

**Comment by tractorcow** [comment]
Yes I see that in other modules too... the only change I can think of was File::exists() changing to include a filesystem check, but can that really be the problem?

**Comment by sb-relaxt-at** [comment]
Changing the inheritance relationship between File and Folder might cause issues in thirdparty code with checks like "$f instanceof File" that might be used to check for a File/Folder, or?

**Comment by tractorcow** [comment]
Yes, well it would have to be done in a major version.

Folders aren't really files, so they shouldn't be treated as such really. They just have some similar properties. :)

**Comment by sb-relaxt-at** [comment]
I just hacked a possible solution for this - yet not fully tested, there might be a better solution at all:

`````
public function updateFilesystem() {
  $filenameBeforeUpdate = $this->Filename;

  parent::updateFilesystem();

  if($filenameBeforeUpdate != $this->Filename){
    // in this case the filename has changed due to a changed parent
    // we need to write the change to the db, otherwise the children won't see it
    $this->write();
    // no need to proceed as this method has already been called again by the write above
    return; 
  }

  // unmodified from here on
  if($this->ID && ($children = $this->AllChildren())) {
    foreach($children as $child) {
      $child->updateFilesystem();
      $child->write(); // might be only necessary if it is a file?
    }
  }
}

**Comment by tractorcow** [comment]
There's also the solution we have in master, which is to have a blank updateFilesystem() for Folder.

https://github.com/silverstripe/silverstripe-framework/blob/master/filesystem/Folder.php#L258

**Comment by sb-relaxt-at** [comment]
There needs to be more magic happening in master (with the new assets storage). Backporting only these lines result in the Folder to be renamed in DB but not in the filesystem (as this renaming happens only in File#updateFilesystem, which is now voided for Folders). Therefore children aren't updated correctly either. Furthermore it would be necessary to move the call to updateFilesystem from File#onAfterWrite to File#onBeforeWrite (as in master) in order to regenerate the filename. So I am not sure which way to go for a PR...

**Comment by tractorcow** [comment]
Yes it is hard. :) Maybe it simply cannot be done in 3.x, and we should focus on getting 4.x working correctly?

**Comment by sb-relaxt-at** [comment]
Well, it is a major issue in our projects, as renaming a folder (with multiple levels of subfolders) results in an error and an incosistent db state. So I would argue for fixing it asap :-). My suggestion from above seems to work well, but I will double test it. I just wanted to know if there are better suggestions to fix this.

**Comment by tractorcow** [comment]
Your fix is probably ok for 3.x. :) Want to knock it into a PR with some tests?

**Comment by AlphaCactus** [comment]
I tested sb-relaxt-at's solution in 3.0.5 and it seems to have fixed the problem for my situation.  In my scenario I was only trying to rename a folder A to B which had a _resampled folder within which would result in folder A and folder A_resampled.  If there were any sub folders that were in the DB then it would throw a rename error and result in duplicated files.

Anyway, seems maybe this has been an issue since 3.0.5?

**Comment by sb-relaxt-at** [comment]
Hm interesting, I thought I couldn't reproduce it in a 3.2.0-beta1. However, just have to add some testcases so that I can make a PR... The fix seems to work well, we are using it in a rather hacky Extension. If someone is interested, until there is a PR for it: [Gist](https://gist.github.com/sb-relaxt-at/daad875fd0c903db9757)
What is the policy about supporting "older" releases? Should we create a PR for 3.1 and 3.2 as well?

**Comment by AlphaCactus** [comment]
This is the code from 3.0.5 if you want to compare:
```
/**
 * Overloaded to call recursively on all contained {@link File} records.
 */
public function updateFilesystem() {
	parent::updateFilesystem();
		// Note: Folders will have been renamed on the filesystem already at this point,
	// File->updateFilesystem() needs to take this into account.
	if($this->ID && ($children = $this->AllChildren())) {
		foreach($children as $child) {
			$child->updateFilesystem();
			$child->write();
		}
	}
}
```

**Comment by sb-relaxt-at** [comment]
Code is the same as in 3.2.0. I just checked it once again, and it perfectly works in 3.2.0-beta1, but it is not working in 3.2.0. Neither @tractorcow nor I could spot a reason for that. Can you confirm that the error occurs in 3.0.5 with the scenario from my first post (1/2/3 -> 1b/2/3)?

**Comment by sb-relaxt-at** [comment]
Okay @tractorcow has been right, the reason is the newly overwritten method [File#exists](https://github.com/silverstripe/silverstripe-framework/compare/3.2.0-beta1...3.2.0#diff-8e70dea638f9f58d5afd42395d40aeb0R229). Checking for parent::exists() causes the issue. I suppose it to be triggered by [File#getRelativePath](https://github.com/silverstripe/silverstripe-framework/blob/3.2.1/filesystem/File.php#L742) which checks whether or not a parent exists. If it does not exist (as during moving the elements), the fallback behaviour is to look for the File directly beneath assets. I believe that the implementations of the exists method and the getRelativePath method are fine. One could argue whether or not the getRelativePath needs to check for existence of the parent Folder, any opinions on that?

**Comment by sb-relaxt-at** [comment]
@AlphaCactus: In 3.0.5 File#exists has not been overwritten, so your issue might be related but needs to have a different reason in the first place.

**Comment by AlphaCactus** [comment]
@sb-relaxt-at ok it does seem that the scenario is different for me. performing your exact steps doesn't appear to result in a problem for me on 3.0.5 (anymore, at least).  In my case I was attempting to rename a folder in code by setting a new Filename on the folder with `setFilename()` and then writing i with `write()`.  Perhaps this isn't the correct way to rename a folder in code, but after all this tinkering around apparently the core code from 3.0.5 does seem to work correctly and I can't reproduce the problem I was having before that caused the error that led to google that led to this issue on github...

**Comment by tractorcow** [comment]
> Okay @tractorcow has been right, the reason is the newly overwritten method File#exists. Checking for parent::exists() causes the issue. 

I think you've found the issue. :) Change `$p->exists()` to `$p->IsInDB()` and voila.

**Comment by tractorcow** [comment]
Re-opened because no one made a fix for it.

I made this at https://github.com/silverstripe/silverstripe-framework/pull/5215

**Comment by tractorcow** [comment]
Co-incidentally, this should fix all the 3.2 regressions in versioned files module.

**Comment by sb-relaxt-at** [comment]
Thanks for fixing it, I overlooked to submit a PR
